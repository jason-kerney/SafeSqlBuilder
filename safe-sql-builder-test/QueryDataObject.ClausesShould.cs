using System.Collections.Generic;
using System.Linq;
using ApprovalTests;
using SafeSqlBuilder;
using SafeSqlBuilder.Generators;
using SafeSqlBuilder.Models;
using Moq;
using Xunit;

namespace SafeSqlBuilder.Tests
{
    public static partial class QueryDataObject
    {
        public class ClausesShould
        {
            // ReSharper disable once InconsistentNaming
            private readonly object IGNORE_ME = "ignore me";

            [Fact]
            public void HaveSingleItemForSimpleFilter()
            {
                const string propName = "PropName";
                var filters = new[]
                {
                    new Filter
                    {
                        Property = propName,
                        includeNulls = true
                    },
                };
                
                var (query, fakeValidColumns, _) = Builder.InitializeItemsForQueryData(new[] {"Field"}, filters: filters);
                var sut = new QueryData(query, fakeValidColumns, new ClauseFactory(fakeValidColumns), "My_Table");

                Assert.Single(sut.Clauses);
                Assert.Single(sut.Clauses.First());
                Assert.Equal(new NullClause($"{propName}_Cleaned").ToString(), sut.Clauses.First().First().ToString());
            }

            [Fact]
            public void ShouldUseClauseFactoryToReturnClauses()
            {
                var expectedFilter1 = new Filter {Property = "Filter1"};
                var expectedFilter2 = new Filter {Property = "Filter2"};
                var expectedFilter3 = new Filter {Property = "Filter3"};

                var fields = new[] {"My_Field"};
                var filters = new[]
                {
                    expectedFilter1,
                    expectedFilter2,
                    expectedFilter3,
                };
                IEnumerable<Clause> ClauseBuilder(Filter f) => new Clause[]
                {
                    new NullClause(f.Property), 
                    new EqualClause(f.Property, IGNORE_ME)
                };
                
                var (query, fakeValidColumns, mockClauseFactory) = Builder.InitializeFakeItemsForQueryData(fields, filters: filters, clauseBuilder: ClauseBuilder);
                var sut = new QueryData(query, fakeValidColumns.Object, mockClauseFactory.Object, "My_Table");

                var actualClauses = sut.Clauses;
                
                Approvals.VerifyAll(actualClauses, clauses => string.Join(", ", clauses.Select(c => $"\"{c}\"")));

                mockClauseFactory
                    .Verify(factory => factory.Build(
                        It.Is<Filter>(f => ReferenceEquals(f, expectedFilter1)))
                    );

                mockClauseFactory
                    .Verify(factory => factory.Build(
                        It.Is<Filter>(f => ReferenceEquals(f, expectedFilter2)))
                    );

                mockClauseFactory
                    .Verify(factory => factory.Build(
                        It.Is<Filter>(f => ReferenceEquals(f, expectedFilter3)))
                    );
            }

            [Fact]
            public void ShouldHandleNullFilters()
            {
                var strings = new[] {"My_Field"};
                IEnumerable<Clause> ClauseBuilder(Filter f) => new Clause[] {new NullClause(f.Property), new EqualClause(f.Property, IGNORE_ME)};

                var (query, fakeValidColumns, fakeClauseFactory) = Builder.InitializeItemsForQueryData(strings, clauseBuilder: ClauseBuilder);

                var sut = new QueryData(query, fakeValidColumns, fakeClauseFactory, "My_Table");

                Assert.Empty(sut.Clauses);
            }
        }
    }
}