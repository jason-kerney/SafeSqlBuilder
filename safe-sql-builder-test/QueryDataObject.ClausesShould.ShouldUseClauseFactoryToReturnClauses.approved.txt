"[Filter1] IS NULL", "[Filter1] = @Filter1_EQUAL"
"[Filter2] IS NULL", "[Filter2] = @Filter2_EQUAL"
"[Filter3] IS NULL", "[Filter3] = @Filter3_EQUAL"
